### PR TITLE
bug: Fix and enhance Gitlab reporter

### LINKED ISSUE #4430: Add missing schema for gitlab-reporter
gitlab-reporter got introduced by #4288 ,
even if there is no official schema, we should define what schema we follow, so we would have BC promise on it and won't break it

### PR DESCRIPTION
**Pull request by ostrolucky**
Also adds a schema file, which fixes #4430

This is according Gitlab support I've got when inquiring why the feature doesn't work and codeclimate spec they follow:
 - https://github.com/codeclimate/platform/blob/master/spec/analyzers/SPEC.md#issues
 - https://docs.gitlab.com/ee/ci/testing/code_quality.html#implement-a-custom-tool

### CODE REVIEW

**Review by kubawerlos — CHANGES_REQUESTED** [review]
Hi @ostrolucky, can you split this to fix and to adding Gitlab?

I'm totally confused why XML is extended together with the new schema.

I always say in such situations that "there is no mention that [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle) does not apply to pull requests" 😉

**Comment by ostrolucky** [comment]
By "enhancing" I mean putting there actual lines instead of hardcoding begin and end to 0 ;) Hence why I think whole PR is a fix. And XML tests are changed because I added actual diff to abstract test case.

**Comment by kubawerlos** [comment]
I saw, I mean all the changes starting from `src/Console/ConfigurationResolver.php` could be in one PR - would be straightforward and easily mergable - something like a fix/enhancement/improving XD.

Adding Gitlab report would be something to discuss more, for me the changes are fine, but I'd like someone who can assure they are the right changes - thus making it separate would make such review easier/faster.

**Comment by ostrolucky** [comment]
WDYM adding gitlab report? I didn't add Gitlab report, I've only added missing schema that I saw was mentioned in #4430. Do you mean that? Gitlab report itself is there since #4288, but is currently broken because Gitlab doesn't parse it.

File: doc/usage.rst [hunk 1]
```diff
@@ -29,6 +29,7 @@ NOTE: the output for the following formats are generated in accordance with sche
 * ``json`` follows the `own JSON schema </doc/schemas/fix/schema.json>`_
 * ``junit`` follows the `JUnit XML schema from Jenkins </doc/schemas/fix/junit-10.xsd>`_
 * ``xml`` follows the `own XML schema </doc/schemas/fix/xml.xsd>`_
+* `gitlab` follows the `codeclimate JSON schema </doc/schemas/fix/codeclimate.json>`_
```
Comment (kubawerlos) [inline doc/usage.rst:32]: Can `gitlab` be surrounded with double backticks as the other entries and placed a few lines higher, so the list keeps its alphabetical order?

File: src/Console/Command/FixCommand.php [hunk 2]
```diff
@@ -100,6 +100,7 @@ public function getHelp(): string
 * `json` follows the `own JSON schema </doc/schemas/fix/schema.json>`_
 * `junit` follows the `JUnit XML schema from Jenkins </doc/schemas/fix/junit-10.xsd>`_
 * `xml` follows the `own XML schema </doc/schemas/fix/xml.xsd>`_
+* `gitlab` follows the `codeclimate JSON schema </doc/schemas/fix/codeclimate.json>`_
```
Comment (kubawerlos) [inline src/Console/Command/FixCommand.php:103]: Can `gitlab` be placed a few lines higher, so the list keeps its alphabetical order?

File: src/Console/Report/FixReport/GitlabReporter.php [hunk 3]
```diff
@@ -39,15 +52,22 @@ public function generate(ReportSummary $reportSummary): string
...
+                    'description' => isset($this->fixers[$fixerName]) ?
+                        $this->fixers[$fixerName]->getDefinition()->getSummary() : '',
+                    'categories' => ['Style'],
                     'fingerprint' => md5($fileName.$fixerName),
                     'severity' => 'minor',
                     'location' => [
                         'path' => $fileName,
                         'lines' => [
-                            'begin' => 0, // line numbers are required in the format, but not available to reports
+                            'begin' => $firstChunk ? $firstChunk->getStart() : 0,
```
Comment (kubawerlos) [inline src/Console/Report/FixReport/GitlabReporter.php:69]: Can the logical condition before `?` be explicitly boolean? Like `$firstChunk instanceof Something`?

File: src/Console/Report/FixReport/ReporterFactory.php [hunk 4]
```diff
@@ -23,11 +24,20 @@
  */
 final class ReporterFactory
 {
+    /** @var array<string, FixerInterface> */
+    private array $fixers;
+
     /**
      * @var array<string, ReporterInterface>
```
Comment (kubawerlos) [inline src/Console/Report/FixReport/ReporterFactory.php:31]: Please make this one single-line or the others multi-line to have them consistent.

File: tests/Test/Assert/AssertJsonSchemaTrait.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+namespace PhpCsFixer\Tests\Test\Assert;
+
+use JsonSchema\Validator;
+
+/** @internal */
+trait AssertJsonSchemaTrait
+{
+    protected static function assertJsonSchema(string $schemaFile, string $json): void
```
Comment (kubawerlos) [inline tests/Test/Assert/AssertJsonSchemaTrait.php:22]: Any reason for change from `private` to `protected`?

**Review by kubawerlos — CHANGES_REQUESTED** [review]

**Review by kubawerlos — APPROVED** [review]

**Comment by ostrolucky** [comment]
When can we expect to get this merged?

File: doc/schemas/fix/codeclimate.json [hunk 6]
```diff
@@ -0,0 +1,97 @@
+{
```
Comment (keradus) [inline doc/schemas/fix/codeclimate.json:1]: can we have the reference where this file is coming from, linked?
so we can refer the original source, but also update the file if anything change in source file

File: doc/usage.rst [hunk 7]
```diff
@@ -26,6 +26,7 @@ The ``--format`` option for the output format. Supported formats are ``txt`` (de
 NOTE: the output for the following formats are generated in accordance with schemas
 
 * ``checkstyle`` follows the common `"checkstyle" XML schema </doc/schemas/fix/checkstyle.xsd>`_
+* ``gitlab`` follows the `codeclimate JSON schema </doc/schemas/fix/codeclimate.json>`_
```
Comment (keradus) [inline doc/usage.rst:29]: this is a new format?
if so, why ticket is marked as bugfix?

Comment (ostrolucky) [reply doc/usage.rst:29]: It's not new, it was missing from usage.rst because there was no schema defined

File: doc/schemas/fix/codeclimate.json (hunk 6 above)
Comment (ostrolucky) [reply doc/schemas/fix/codeclimate.json:1]: Sure, how do you suggest to link it?

File: doc/usage.rst (hunk 7 above)
Comment (keradus) [reply doc/usage.rst:29]: got you!
can we put a test for this file content, that ensures the all reporters are present?

File: doc/schemas/fix/codeclimate.json (hunk 6 above)
Comment (keradus) [reply doc/schemas/fix/codeclimate.json:1]: if possible in this file directly, then in this file.
if not, as a note in folder, (like codeclimate-ref.md)

File: src/Console/Report/FixReport/ReporterFactory.php [hunk 8]
```diff
@@ -48,7 +56,7 @@ public function registerBuiltInReporters(): self
         }
 
         foreach ($builtInReporters as $reporterClass) {
-            $this->registerReporter(new $reporterClass());
+            $this->registerReporter(new $reporterClass($this->fixers));
```
Comment (keradus) [inline src/Console/Report/FixReport/ReporterFactory.php:59]: why this change?

Comment (ostrolucky) [reply src/Console/Report/FixReport/ReporterFactory.php:59]: To be able to get description of a fixer here https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7089/files#diff-96b8f6fe35684550b46a932958362784f237a3fe0135d7aa8d71421868ac5999R63

File: doc/usage.rst (hunk 7 above)
Comment (ostrolucky) [reply doc/usage.rst:29]: It was done, see here ;) https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7089/files#diff-4b42eca6a4843666251d64385cf19ef941b679ff4fa63695ff2861d66bcc1846R163

Comment (keradus) [reply doc/usage.rst:29]: I meant not a test that gitlab-report output is matching the spec file, but that all possible report formats are enumareted in `doc/usage.rst` (like, you had to add new line - it should not be possible to forget adding the new report class to the docs and if one would, test should fail)

File: src/Console/Report/FixReport/ReporterFactory.php (hunk 8 above)
Comment (keradus) [reply src/Console/Report/FixReport/ReporterFactory.php:59]: I think it's very big dependency for quite small usage.
I also wonder if we need whole FixerInterface functionality, or we only pass definitions (or even only descriptions)

File: doc/usage.rst (hunk 7 above)
Comment (keradus) [reply doc/usage.rst:29]: i hoped for scout rule ;)

File: src/Console/Report/FixReport/ReporterFactory.php (hunk 8 above)
Comment (ostrolucky) [reply src/Console/Report/FixReport/ReporterFactory.php:59]: Without description it would look like this in Gitlab. Notice that Psalm has better description, because it exposes actual description instead of only rule name

image

As for why Gitlab only: Because I'm not familiar with other reporters. Gitlab is something I use only at the moment for this. If somebody want to use it different formatter, after my change it will be straightforward now :)

Anyways I don't want to force this change, I am just trying to have a best result. If you say it can't be done then fine.

File: doc/usage.rst (hunk 7 above)
Comment (ostrolucky) [reply doc/usage.rst:29]: I didn't even plan to add a schema file, but I decided since I'm already fixing Gitlab formatter I added it. Please don't punish me for that and add more and more on top of it 😅

File: src/Console/Report/FixReport/ReporterFactory.php (hunk 8 above)
Comment (Wirone) [reply src/Console/Report/FixReport/ReporterFactory.php:59]: @ostrolucky can you share how it looks like with the description, so we can compare it visually?

Comment (ostrolucky) [reply src/Console/Report/FixReport/ReporterFactory.php:59]: This is an approximation, but I think it will look like this
image

Comment (Wirone) [reply src/Console/Report/FixReport/ReporterFactory.php:59]: Personally I prefer version **without** description. It makes much more sense to display error message if the tool reports it (like from Psalm or PHPStan), but fixer's description is misleading here - it's neither an error message, nor proper requirement. Some descriptions may be longer too, so considering multiple occurrences, it would make such output really messy. I believe it would be better with fixer's name only (if possible, with link to `https://cs.symfony.com/doc/rules/basic/<NAME>.html`, but only for built-in fixers).

Comment (ostrolucky) [reply src/Console/Report/FixReport/ReporterFactory.php:59]: I agree. I will make it the first one then.

**Review by Wirone — APPROVED** [review]

File: doc/usage.rst (hunk 7 above)
Comment (keradus) [reply doc/usage.rst:29]: let's at least create a follow-up issue for that missing test

Comment (Wirone) [reply doc/usage.rst:29]: Created #7106.

**Review by keradus — APPROVED** [review]

**Comment by Wirone** [comment]
Thank you very much @ostrolucky 🍻
